### ISSUE TITLE
bugs in pdf_document extra_dependencies

### ISSUE THREAD

**Issue by haozhu233** [rstudio/rmarkdown#910]
Hi, I discovered a few bugs in the `extra_dependencies` option in `pdf_document`. I found them out in an afford trying to load a latex package from a R function while using rmarkdown + knitr. 

1. [Fixed in PR #911 ] First of all, in the current github version of rmarkdown (https://github.com/rstudio/rmarkdown/blob/master/R/pdf_document.R#L176-L187), for `pdf_documents`, the content of `extra_dependency` is passed to a list called `format_deps`. This `format_deps` will not be put into real use unless `has_latex_dependencies(knit_meta)` is `TRUE`.  I think this bug will prevent people from using `extra_dependency` in most cases. 

~~2. Also, if people use the results of `latex_dependency()` directly to `extra_dependency`, they will receive the following error message:~~
~~YAML header:~~
```
---
title: "Untitled"
output: 
  pdf_document:
    keep_tex: true
    extra_dependencies: "`r rmarkdown::latex_dependency('multirow')`"
---
```
~~Error:~~
```
processing file: Untitled.Rmd
Quitting from lines 2-8 (Untitled.Rmd) 
Error in vapply(x, format_sci_one, character(1L), ..., USE.NAMES = FALSE) : 
  values must be length 1,
 but FUN(X[[2]]) result is length 0
Calls: <Anonymous> ... paste -> hook -> .inline.hook -> format_sci -> vapply
Execution halted
```

3. [Fixed in PR #911 ] Even if people find a way around to get the results of `latex_dependency` to `pdf_document`, the results of `latex_dependency`, as a list of 3, should not be append to `format_deps` directly (https://github.com/rstudio/rmarkdown/blob/master/R/pdf_document.R#L177). We should put a list round it so the whole list can be passed as an element. 

I have worked around solutions to question 1 & 3 and I will submit a PR very soon. Also, I added a feature so people can use `extra_dependencies` in the YAML section more easily.

**Comment by yihui** [comment]
For 2, instead of `` `r ` ``, you should use `!expr `. I know it is confusing, but `` `r ` `` is knitr's syntax for including inline output (typically it is supposed to be a character string). If you want to return an R value that is passed to the `pdf_document()` function, you need to use the yaml syntax, which is `!expr` followed by R code to generate an arbitrary value.

**Comment by haozhu233** [comment]
@yihui 👍 Thanks for letting me know that! Now my problem is solved. :)

**Comment by yihui** [comment]
You don't need to (and it is better not to) close this issue or the PR #911. They will be automatically closed after #944 is merged; before then, the bugs you reported still exist.
